Mathlib/Algebra/Algebra/Subalgebra/Basic.lean

```diff
@@ -938,6 +938,20 @@ variable (f : A →ₐ[R] B)
 theorem range_comp_val : (f.comp S.val).range = S.map f := by
   rw [AlgHom.range_comp, range_val]
 
+/-- An `AlgHom` between two rings restricts to an `AlgHom` from any subalgebra of the
+domain onto the image of that subalgebra. -/
+def _root_.AlgHom.subalgebraMap : S →ₐ[R] S.map f :=
+  (equivOfEq _ _ (range_comp_val S f)).toAlgHom.comp (f.comp S.val).rangeRestrict
```

Comment:
Parallel to `LinearMap.submoduleMap`, I suggest you use [AlgHom.codRestrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#AlgHom.codRestrict) to define this. Hopefully the proof of the last lemma becomes easier. (Can't you copy the same proof now?)

Reply:
> Parallel to `LinearMap.submoduleMap`, I suggest you use [AlgHom.codRestrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#AlgHom.codRestrict) to define this. Hopefully the proof of the last lemma becomes easier. (Can't you copy the same proof now?)

I'm not sure, since last lemma says `Surjective`. `rangeRestrict` is naturally surjective ([AlgHom.rangeRestrict_surjective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#AlgHom.rangeRestrict_surjective)) but we don't have this for `codRestrict`. Let me think a bit.

Reply:
> Can't you copy the same proof now?

No, as the definition of `LinearMap.submoduleMap` uses [LinearMap.restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/LinearMap.html#LinearMap.restrict) which is not available for `AlgHom`...

I think the last proof becomes one line if we have `AlgEquiv.comp_surjective`, but seems that it's only for `EquivLike` which provided a unified API but each time we need `simp only XXX`.

Maybe I can try `simpa using XXX`...

Reply:
OK `simpa` works. Do you think this is OK?

Reply:
I was suggesting
```suggestion
def _root_.AlgHom.subalgebraMap : S →ₐ[R] S.map f :=
  (f.comp S.val).codRestrict _ fun x ↦ ⟨_, x.2, rfl⟩
```